pkg/build/legacy/parser/deffile.go

```diff
@@ -0,0 +1,413 @@
+//
+// If there are any Golang devs reading this, please improve your documentation for this. It's awful.
+func scanDefinitionFile(data []byte, atEOF bool) (advance int, token []byte, err error) {
+	inSection := false
+	var retbuf bytes.Buffer
+	advance = 0
+
+	l := len(data)
+
+	for advance < l {
+		// We are essentially a pretty wrapper to bufio.ScanLines.
+		a, line, err := bufio.ScanLines(data[advance:], atEOF)
+		if err != nil && err != bufio.ErrFinalToken {
+			return 0, nil, err
+		} else if line == nil { // If ScanLines returns a nil line, it needs more data. Send req for more data
+			return 0, nil, nil // Returning 0, nil, nil requests Scanner.Scan() method find more data or EOF
+		}
+
+		_, word, err := bufio.ScanWords(line, true) // Tokenize the line into words
+		if err != nil && err != bufio.ErrFinalToken {
+			return 0, nil, err
+		}
+
+		// Check if the first word starts with % sign
+		if word != nil && word[0] == '%' {
+			// If the word starts with %, it's a section identifier
+
+			// We no longer check if the word is a valid section identifier here, since we want to move to
+			// a more modular approach where we can parse arbitrary sections
+			if inSection {
+				// Here we found the end of the section
+				return advance, retbuf.Bytes(), nil
+			} else if advance == 0 {
+				// When advance == 0 and we found a section identifier, that means we have already
+				// parsed the header out and left the % as the first character in the data. This means
+				// we can now parse into sections.
+				retbuf.Write(line)
+				retbuf.WriteString("\n")
+				inSection = true
+			} else {
+				// When advance != 0, that means we found the start of a section but there is
+				// data before it. We return the data up to the first % and that is the header
+				return advance, retbuf.Bytes(), nil
+			}
+		} else {
+			// This line is not a section identifier
+			retbuf.Write(line)
+			retbuf.WriteString("\n")
+		}
+
+		// Shift the advance retval to the next line
+		advance += a
+		if a == 0 {
+			break
+		}
+	}
+
+	if !atEOF {
+		return 0, nil, nil
+	}
+
+	return advance, retbuf.Bytes(), nil
+
+}
+
+func getSectionName(line string) string {
+	// trim % prefix on section name
+	line = strings.TrimLeft(line, "%")
+	lineSplit := strings.SplitN(strings.ToLower(line), " ", 2)
+
+	return lineSplit[0]
+}
+
+// parseTokenSection splits the token into maximum 2 strings separated by a newline,
+// and then inserts the section into the sections map
+//
+func parseTokenSection(tok string, sections map[string]string) error {
+	split := strings.SplitN(tok, "\n", 2)
+	if len(split) != 2 {
+		return fmt.Errorf("Section %v: Could not be split into section name and body", split[0])
+	}
+
+	key := getSectionName(split[0])
+	if appSections[key] {
+		sectionSplit := strings.SplitN(strings.TrimLeft(split[0], "%"), " ", 3)
+		if len(sectionSplit) < 2 {
+			return fmt.Errorf("App Section %v: Could not be split into section name and app name", sectionSplit[0])
```

Comment:
```suggestion
			return fmt.Errorf("app section %v: could not be split into section name and app name", sectionSplit[0])
```

Reply:
How did CCI/Travis pass with the caps in the error message? I thought it was a requirement that theres no caps in error messages?...

Reply:
I don't think it's a hard requirement, just good practice.